### PR TITLE
Refactor authorization

### PR DESCRIPTION
**Pull request by jannyHou**
Some refactor for https://github.com/strongloop/loopback-next/pull/1205

It contains subsequent changes to make the e2e tests in PR https://github.com/strongloop/loopback4-example-shopping/pull/231 work.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by jannyHou** [comment]
Tests pass on my local. The CI error is same as https://github.com/strongloop/loopback-next/pull/3520

File: packages/authorization/src/__tests__/acceptance/authorization-casbin.acceptance.ts [hunk 1]
```diff
@@ -26,7 +27,7 @@ describe('Authorization', () => {
   let events: string[];
 
   before(givenApplicationAndAuthorizer);
-  beforeEach(givenRequestContext);
+  beforeEach(() => givenRequestContext());
```
Comment (raymondfeng) [inline packages/authorization/src/__tests__/acceptance/authorization-casbin.acceptance.ts]: Why is this change needed?

File: packages/authorization/src/__tests__/acceptance/authorization-casbin.acceptance.ts [hunk 2]
```diff
@@ -161,6 +165,7 @@ describe('Authorization', () => {
       __dirname,
       '../../../fixtures/casbin/rbac_policy.csv',
     );
-    return await casbin.newEnforcer(conf, policy);
+    const result = await casbin.newEnforcer(conf, policy);
+    return result;
```
Comment (raymondfeng) [inline packages/authorization/src/__tests__/acceptance/authorization-casbin.acceptance.ts]: `return casbin.newEnforcer(conf, policy);` should be good enough.

File: packages/authorization/src/authorize-interceptor.ts [hunk 3]
```diff
@@ -41,12 +47,19 @@ export class AuthorizationInterceptor implements Provider<Interceptor> {
...
+      const result = await next();
+      return result;
     }
     debug('Authorization metadata for %s', description, metadata);
-    const user = await invocationCtx.get<{name: string}>('current.user', {
-      optional: true,
-    });
+
+    // retrieve it from authentication module
+    const user = await invocationCtx.get<{name: string}>(
```
Comment (raymondfeng) [inline packages/authorization/src/authorize-interceptor.ts]: Use `UserProfile` instead of `{name: string}`?

File: packages/authorization/src/__tests__/acceptance/authorization-casbin.acceptance.ts (hunk 1 above)
Comment (jannyHou) [reply packages/authorization/src/__tests__/acceptance/authorization-casbin.acceptance.ts]: @raymondfeng I am seeing this weird error...

Screen Shot 2019-08-09 at 3 29 35 PM

File: packages/authorization/src/authorize-interceptor.ts [hunk 4]
```diff
@@ -55,19 +68,27 @@ export class AuthorizationInterceptor implements Provider<Interceptor> {
...
     };
+
     debug('Security context for %s', description, authorizationCtx);
     let authorizers = await loadAuthorizers(
       invocationCtx,
       metadata.voters || [],
     );
+
+    // pass `currentUser` to authorizers
+    Object.assign(metadata, {currentUser: user});
```
Comment (raymondfeng) [inline packages/authorization/src/authorize-interceptor.ts]: `currentUser` should be set to `authorizationContext` if necessary instead of `metadata`.

File: packages/authorization/src/authorize-interceptor.ts [hunk 5]
```diff
@@ -55,19 +68,27 @@ export class AuthorizationInterceptor implements Provider<Interceptor> {
...
+      // we can add another interceptor to process the error
       if (decision === AuthorizationDecision.DENY) {
-        throw new Error('Access denied');
+        const error = new AuthorizationError('Access denied');
+        error.statusCode = 401;
+        throw error;
       }
     }
-    return await next();
+    const result = await next();
```
Comment (raymondfeng) [inline packages/authorization/src/authorize-interceptor.ts]: `return next();` should be good.

File: packages/authorization/package.json [hunk 6]
```diff
@@ -27,6 +27,7 @@
   "devDependencies": {
     "@loopback/build": "^2.0.3",
     "@loopback/testlab": "^1.6.3",
+    "@loopback/authentication": "^2.1.8",
```
Comment (raymondfeng) [inline packages/authorization/package.json]: Please add some comments about this temporary workaround.

Comment (jannyHou) [reply packages/authorization/package.json]: @raymondfeng I added an entry in the README.md file to explain the user injection.

**Comment by jannyHou** [comment]
https://github.com/strongloop/loopback4-example-shopping/pull/231 is also updated.

File: packages/authorization/src/authorize-interceptor.ts (hunk 4 above)
Comment (jannyHou) [reply packages/authorization/src/authorize-interceptor.ts]: https://github.com/strongloop/loopback-next/pull/3529 is also updated accordingly.

File: packages/authorization/src/__tests__/acceptance/authorization-casbin.acceptance.ts (hunk 1 above)
Comment (raymondfeng) [reply packages/authorization/src/__tests__/acceptance/authorization-casbin.acceptance.ts]: Oh, I think that's because `givenRequestContext` takes an optional `user` argument.

File: packages/authorization/src/authorize-interceptor.ts [hunk 7]
```diff
@@ -41,33 +48,55 @@ export class AuthorizationInterceptor implements Provider<Interceptor> {
...
+    // This is a workaround before we extract a common layer
+    // for authentication and authorization.
+    function userToPrinciple(user: UserProfile): Principal {
+      return {
+        name: user.name || user.id,
+        id: user.id,
+        email: user.email,
+        type: 'USER'
+      }
+    }
```
Comment (raymondfeng) [inline packages/authorization/src/authorize-interceptor.ts]: Please move this inline function to the module scope.

File: packages/authorization/package.json (hunk 6 above)
Comment (raymondfeng) [reply packages/authorization/package.json]: BTW, let's change the initial version to be `0.0.1` so that the first release will be `0.1.0`.
